src/core/ComboBox/ComboBox.tsx

```diff
@@ -380,84 +454,363 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+    return (
+      <span
+        ref={refs}
+        className={cx('iui-end-icon', className)}
+        onClick={(e) => {
+          onClickProp?.(e);
+          if (!e.isDefaultPrevented()) {
+            dispatch(['toggle']);
+          }
+        }}
+        {...rest}
+      >
+        {children ?? <SvgCaretDownSmall aria-hidden />}
+      </span>
+    );
+  },
+);
+ComboBoxEndIcon.displayName = 'ComboBoxEndIcon';
+
+const ComboBoxInput = React.forwardRef(
+  (props: InputProps, forwardedRef: React.RefObject<HTMLInputElement>) => {
+    const { onKeyDown: onKeyDownProp, onFocus: onFocusProp, ...rest } = props;
+
+    const { isOpen, id, focusedIndex } = useSafeContext(ComboBoxStateContext);
+    const dispatch = useSafeContext(ComboBoxActionContext);
+    const { inputRef, menuRef, optionsRef } = useSafeContext(
+      ComboBoxRefsContext,
+    );
+    const refs = useMergedRefs(inputRef, forwardedRef);
+
+    const focusedIndexRef = React.useRef(focusedIndex ?? -1);
+    React.useEffect(() => {
+      focusedIndexRef.current = focusedIndex ?? -1;
+    }, [focusedIndex]);
+
+    const getIdFromIndex = (index: number) => {
+      return (
+        menuRef.current?.querySelector(`[data-iui-index="${index}"]`)?.id ?? ''
+      );
+    };
+
+    const handleKeyDown = React.useCallback(
+      (event: React.KeyboardEvent<HTMLInputElement>) => {
+        onKeyDownProp?.(event);
+        const length = optionsRef.current.length ?? 0;
+
+        switch (event.key) {
+          case 'ArrowDown': {
+            event.preventDefault();
+            if (!isOpen) {
+              dispatch(['open']);
+            }
+
+            if (length === 0) {
+              return;
+            }
+
+            if (focusedIndexRef.current === -1) {
+              return dispatch([
+                'focus',
+                Number(optionsRef.current[0].dataset.iuiIndex),
+              ]);
             }
-          }}
-          animation='shift-away'
-          duration={200}
-          {...dropdownMenuProps}
-          content={
-            <Menu
-              id={`${id}-list`}
-              className='iui-scroll'
-              style={{
-                minWidth,
-                maxWidth: `min(${minWidth * 2}px, 90vw)`,
-                maxHeight: 300,
-              }}
-              setFocus={false}
-              role='listbox'
-              ref={menuRef}
-            >
-              {menuItems}
-            </Menu>
+
+            let nextIndex = focusedIndexRef.current;
+            do {
+              const currentElement = menuRef.current?.querySelector(
+                `[data-iui-index="${nextIndex}"]`,
+              );
+              const nextElement =
+                currentElement?.nextElementSibling ??
+                menuRef.current?.querySelector('[data-iui-index]');
+              nextIndex = Number(nextElement?.getAttribute('data-iui-index'));
+              const __originalIndex = Number(
```

Comment:
amm.. why you need `__originalIndex` ? Value same as `nextIndex` just with -1 if get undefined attribute.

Reply:
removed